spec/mem/tree.md

```diff
@@ -0,0 +1,564 @@
+        }
+    }
+}
+```
+
+We define the tree structure to track reborrows in each allocation. The tree consists of several nodes, each corresponding to a borrow tag (i.e., a reborrow). Each node has a parent and a list of children. Additionally, we track each tag's permissions on memory locations (bytes) within the allocation.
+
+```rust
+pub struct Tree {
+    root_tag: BorTag, 
+    /// Map a borrow tag to a node
+    nodes: Map<BorTag, Node>,
+}
+
+pub struct Node {
+    /// Borrow tag of the parent node
+    parent: Option<BorTag>, 
+    /// Borrow tags of the children node
+    children: List<BorTag>,
+    /// Permission for each location
+    permissions: Map<Address, Permission>,
+}
+
+pub enum NodeRelation {
+    Child, 
+    Foreign,
+}
+
+impl Tree {
+    /// Insert a node into the tree
+    fn insert_node(&mut self, tag: BorTag, node: Node) {
+        self.nodes.insert(tag, node);
+    }
+}
+```
+
+To update the tree during each memory access, each node in the tree can be divided into two disjoint sets: Child and Foreign. The Child set includes the node itself and all its descendants, while the Foreign set contains all other nodes. When a node is accessed, we must update and check permissions for each node in the tree based on its relative position to the accessed node.
+
+```rust
+impl TreeBorrowsAllocation {
+    /// Recusively do permission transition on the `curr` node and all its descendants.
+    /// `base` means the node that is actually memory accessed, and the `node_relation` means the 
+    /// relation between `curr` and `base` from the perspective of `curr`.
+    /// For instance, when the `curr` is one of descendants of the `base`,
+    /// the `node_relation` should be `NodeRelation::Foreign`.
+    /// This method will throw UBs, representing the undefined behavior captured by Tree Borrows.
+    fn node_access(
+        &mut self, 
+        curr: BorTag,
+        access_pattern: AccessKind,
+        ptr: Pointer<TreeBorrowsProvenance>,
+        size: Size,
+        base: BorTag,
+        node_relation: NodeRelation
+    ) -> Result {
+        let Some(mut node) = self.tree.nodes.get(curr) else { throw_ub!("Tree Borrows: node not existed"); };
+
+        // Update the permission for each location
+        for location in ptr.addr..ptr.addr + size.bytes() {
+            let Some(curr_permission) = node.permissions.get(location) else {
+                throw_ub!("Tree Borrows: location without permission");
+            };
+
+            let next_permission = Self::permission_transition(curr_permission, access_pattern, node_relation)?;
+            node.permissions.insert(location, next_permission);
+        }
+
+        // Insert back the updated node
+        self.tree.nodes.insert(curr, node);
+
+        // For each descendant, the base node is a foreign.
+        let new_relation = if curr == base { NodeRelation::Foreign } else { node_relation };
```

Comment:
This is actually incorrect. It can not differ strict transitive parents of `base` and nodes who do not share a branch with `base`